pyluna-pathology/luna/pathology/cli/extract_shape_features.py

```diff
@@ -0,0 +1,173 @@
+            -o ./shape_features.csv
+
+    """
+    cli_runner(cli_kwargs, _params_, extract_shape_features)
+
+
+def extract_shape_features(
+    input_slide_mask: str, label_cols: List[str], output_dir: str
+):
+    """Extracts shape and spatial features (HIF features) from a slide mask
+
+     Args:
+        input_slide_mask (str): path to slide mask (*.tif)
+        label_cols (List[str]): list of labels that coorespond to those in input_slide_tiles
+        output_dir (str): output/working directory
+
+    Returns:
+        dict: output .tif path and the number of shapes for which features were generated 
+    """
+    # List of features to extract.
+    # Default behavior of regionprops_table only generates label and bbox features.
+    # Not all of these features may be relevant 
+    properties = [
+        "area",
+        "bbox",
+        "bbox_area",
+        "centroid",
+        "convex_area",
+        "convex_image",
+        "coords",
+        "eccentricity",
+        "equivalent_diameter",
+        "euler_number",
+        "extent",
+        "filled_area",
+        "filled_image",
+        "image",
+        "inertia_tensor",
+        "inertia_tensor_eigvals",
+        "label",
+        "local_centroid",
+        "major_axis_length",
+        "minor_axis_length",
+        "moments",
+        "moments_central",
+        "moments_hu",
+        "moments_normalized",
+        "orientation",
+        "perimeter",
+        "slice",
+        "solidity",
+    ]
+
+    mask = tifffile.imread(input_slide_mask)
+    logger.info(f"Mask shape={mask.shape}")
+
+    mask_values = {k: v + 1 for v, k in enumerate(label_cols)}
+    logger.info(f"Mapping mask values to labels: {mask_values}")
+    label_mapper = {v: k for k, v in mask_values.items()}
+
+    logger.info("Extracting whole slide features")
+    # gathering whole slide features, one vector per label
+    whole_slide_features = measure.regionprops_table(
+        label_image=mask, properties=properties
+    )
+    whole_slide_features_df = pd.DataFrame.from_dict(whole_slide_features)
+
+    # add column with label name
+    whole_slide_features_df["label_name"] = "whole_" + whole_slide_features_df[
+        "label"
+    ].map(label_mapper)
+    logger.info(
+        f"Extracted whole slide features for {len(whole_slide_features_df)} labels"
+    )
+
+    logger.info("Extracting regional features based on connectivity")
+    mask_label = measure.label(mask, connectivity=2)
+    # pass intensity image to propogate label type
+    properties.extend(
+        [
+            "min_intensity",
+            "max_intensity",
+        ]
+    )
+    regional_features = measure.regionprops_table(
+        label_image=mask_label, intensity_image=mask, properties=properties
+    )
+    regional_features_df = pd.DataFrame.from_dict(regional_features)
+
+    # verify intensity matches label value
+    assert regional_features["min_intensity"] == regional_features["max_intensity"]
```

Comment:
If it's bad (and possible) for these two to not match, raise a nice error message instead.

Reply:
Yeah I can remove this, there should never be a case where they don't match. I included this because extracting the regional features is requires a slight abuse of the `mask_label` parameter.

In order to extract these features, you need to generate a different mask from the one generated by the previous CLI tool. The skimage function used  `measure.label(mask, connectivity=2)` generates a new mask with unique values for each connected region, which removes the tissue type information previously encoded by the mask. So, I pass the original mask as an intensity image to ensure that that each region can be associated with a tissue type. So each region should have the same `min_intensity` and `max_intensity`, since each region has an intensity given by its tissue type, these intensity values don't relate at all to the WSI underlying all of this. 

I can make this more clear in the documentation.